caffe2/serialize/inline_container.h

```diff
@@ -109,15 +112,16 @@ class CAFFE2_API PyTorchStreamReader final {
   ~PyTorchStreamReader();
 
  private:
-   size_t read(uint64_t pos, char* buf, size_t n);
-   void valid(const char* what);
-   size_t getFileID(const std::string& name);
-
-   friend size_t istream_read_func(void *pOpaque, uint64_t file_ofs, void *pBuf, size_t n);
-   std::unique_ptr<mz_zip_archive> ar_;
-   std::string archive_name_;
-   std::istream* in_;
-   std::ifstream file_stream_;
+  void init();
+  size_t read(uint64_t pos, char* buf, size_t n);
+  void valid(const char* what);
+  size_t getFileID(const std::string& name);
+
+  friend size_t istream_read_func(void *pOpaque, uint64_t file_ofs, void *pBuf, size_t n);
+  std::unique_ptr<mz_zip_archive> ar_;
+  std::string archive_name_;
+  std::shared_ptr<at::ReadAdapterInterface> in_;
+  std::ifstream file_stream_;
```

Comment:
Now that there is a third way to access the reader, it is messy to have ifstream initialized the way it is. It would make more sense to have a FileReader subclass of ReaderAdapterInterface of the reader interface be responsible for opening, reading, and closing the file if that constructor ran.